src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/MessageBody.cs

```diff
@@ -291,15 +298,20 @@ protected virtual void OnConsumedBytes(int count)
                 }
             }
 
+            if (upgrade)
+            {
+                return new ForUpgrade(context);
+            }
+
             return keepAlive ? _zeroContentLengthKeepAlive : _zeroContentLengthClose;
         }
 
-        private class ForRemainingData : MessageBody
+        private class ForUpgrade : MessageBody
         {
-            public ForRemainingData(bool upgrade, Frame context)
+            public ForUpgrade(Frame context)
                 : base(context)
             {
-                RequestUpgrade = upgrade;
+                RequestUpgrade = true;
             }
 
             protected override ValueTask<ArraySegment<byte>> PeekAsync(CancellationToken cancellationToken)
```

Comment:
Shouldn't this always return `new ValueTask<ArraySegment<byte>>();` since the upgrade requests can't have a payload? I think doing otherwise is inviting buffering middleware to interfere.